### DISCUSSION TITLE
How to extract more than one tables available in the single page pdf as separate  data frame or md or csv?

### DISCUSSION THREAD

**Discussion by thangarajdeivasikamani** [pymupdf/PyMuPDF-Utilities#138] [Q&A]
Hello Team,

I have the pdf which has more than one tables in single page.
Example :

First table have 5 columns and 10 rows, 2nd Tables have 10 column,3 rows , 3rd Table  3 columns , 20 rows.

In that case how to extract the each table as individual data frame and  combine at page level.

**Comment by JorjMcKie** [comment]
Create an empty list to be used as a list of data frames. Then, for each table encountered, append `table.to_pandas()` to that list.
Then look up the pandas documentation to see what you can do with separate data frames.

**Reply by JorjMcKie** [reply]
For example, pandas offers joining data frames, or outputting them to CSV, Excel or also markdown.

**Reply by thangarajdeivasikamani** [reply]
if my pdf has single Table the below code able to find the Table and creating the data frame.


_[tabs = page.find_tables()  # detect the tables
print(tabs)
for i,tab in enumerate(tabs):  # iterate over all tables
    print(i,tab)
    for cell in tab.header.cells:
        print(cell)
        page.draw_rect(cell,color=fitz.pdfcolor["red"],width=0.3)
        print("--")
    page.draw_rect(tab.bbox,color=fitz.pdfcolor["green"])
    print(f"Table {i} column names: {tab.header.names}, external: {tab.header.external}")
    
show_image(page, f"Table & Header BBoxes")]_

output

pymupdf.table.TableFinder object at 0x000002A8F9B09590
0 pymupdf.table.Table object at 0x000002A8EA7C2810
(308.22000885009766, 631.5, 380.85880317962426, 649.6500244140625)
--
(380.85880317962426, 631.5, 528.3300018310547, 649.6500244140625)
--
Table 0 column names: ['Reference', 'Part number'], external: False


But the pdf page has multiple Tables the same code giving the below error. Not sure what could be wrong.

output

pymupdf.table.TableFinder object at 0x000002A8FADDF410
0 pymupdf.table.Table object at 0x000002A8EB257B10
(39.20347250075567, 151.95047505696616, 152.50390625, 165.32421875)
--
None

Error Log:

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~\AppData\Local\Temp\ipykernel_30544\925744604.py in ?()
      3 for i,tab in enumerate(tabs):  # iterate over all tables
      4     print(i,tab)
      5     for cell in tab.header.cells:
      6         print(cell)
----> 7         page.draw_rect(cell,color=fitz.pdfcolor["red"],width=0.3)
      8         print("--")
      9     page.draw_rect(tab.bbox,color=fitz.pdfcolor["green"])
     10     print(f"Table {i} column names: {tab.header.names}, external: {tab.header.external}")

~\.conda\envs\oolamaexp\Lib\site-packages\pymupdf\utils.py in ?(page, rect, color, fill, dashes, width, lineCap, lineJoin, morph, overlay, stroke_opacity, fill_opacity, oc, radius)
   2159     '''
   2160     Draw a rectangle. See Shape class method for details.
   2161     '''
   2162     img = page.new_shape()
-> 2163     Q = img.draw_rect(pymupdf.Rect(rect), radius=radius)
   2164     img.finish(
   2165         color=color,
   2166         fill=fill,

~\.conda\envs\oolamaexp\Lib\site-packages\pymupdf\__init__.py in ?(self, p0, p1, x0, y0, x1, y1, *args)
  11004 
  11005         Explicit keyword args p0, p1, x0, y0, x1, y1 override earlier settings
  11006         if not None.
  11007         """
> 11008         x0, y0, x1, y1 = util_make_rect( *args, p0=p0, p1=p1, x0=x0, y0=y0, x1=x1, y1=y1)
  11009         self.x0 = float( x0)
  11010         self.y0 = float( y0)
  11011         self.x1 = float( x1)

~\.conda\envs\oolamaexp\Lib\site-packages\pymupdf\__init__.py in ?(p0, p1, x0, y0, x1, y1, *args)
  20234                 return args[0], args[1], x1, y1
  20235         elif len(args) == 4:
  20236             return args[0], args[1], args[2], args[3]
  20237         raise Exception( f'Unrecognised args: {args}')
> 20238     ret_x0, ret_y0, ret_x1, ret_y1 = handle_args()
  20239     if p0 is not None:  ret_x0, ret_y0 = get_xy(p0)
  20240     if p1 is not None:  ret_x1, ret_y1 = get_xy(p1)
  20241     if x0 is not None:  ret_x0 = x0

ValueError: not enough values to unpack (expected 4, got 1)

**Reply by thangarajdeivasikamani** [reply]
After comment the below line. It's extract all the tables from respective page.

tabs = page.find_tables()  # detect the tables
#print(tabs)
for i,tab in enumerate(tabs):  # iterate over all tables
    #print(i,tab)
    for cell in tab.header.cells:
        print(cell)
        #page.draw_rect(cell,color=fitz.pdfcolor["red"],width=0.3)
    page.draw_rect(tab.bbox,color=fitz.pdfcolor["green"])
    print(tab.bbox)
    print(f"Table {i} column names: {tab.header.names}, external: {tab.header.external}")
    print("First table")
show_image(page, f"Table & Header BBoxes")


Is there any option to save the image as file instead of  show_image in the last line.
